### PR TITLE
ls --dired: adjust our code after GNU v9.5

### PR DESCRIPTION
**Pull request by sylvestre**
regressed because of https://github.com/uutils/coreutils/pull/6139

### CODE REVIEW

**Comment by cakebaker** [comment]
Did you see that `test_ls_dired_hyperlink_disabled` fails on Windows?

File: src/uu/ls/src/ls.rs [hunk 1]
```diff
@@ -1084,7 +1079,8 @@ impl Config {
 
         let dired = options.get_flag(options::DIRED);
         if dired && format != Format::Long {
-            return Err(Box::new(LsError::ConflictingArgumentDired));
+            // --dired implies --format=long
+            format = Format::Long;
         }
```
Comment (BenWiederhake) [inline src/uu/ls/src/ls.rs:1084]: Nope, it appears that GNU ls has some different semantics:
```console
$ ls --version
ls (GNU coreutils) 9.4
[…]
$ ls --dired                         
CODE_OF_CONDUCT.md  CONTRIBUTING.md  Cargo.toml      GNUmakefile  Makefile       README.md  b      build.rs  deny.toml  foo   fuzz      oranda.json    showargs  target  util
COMMANDS.txt        Cargo.lock       DEVELOPMENT.md  LICENSE      Makefile.toml  a          build  check.sh  docs       foo2  often.sh  renovate.json  src       tests
$ ls --dired --format=single-column
CODE_OF_CONDUCT.md
COMMANDS.txt
CONTRIBUTING.md
Cargo.lock
Cargo.toml
DEVELOPMENT.md
GNUmakefile
LICENSE
Makefile
Makefile.toml
README.md
a
b
build
build.rs
check.sh
deny.toml
docs
[… more lines …]
```

Perhaps it only sets the default value?

**Review by BenWiederhake — CHANGES_REQUESTED** [review]

File: src/uu/ls/src/ls.rs (hunk 1 above)
Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1084]: well, i worked on the GNU implementation (with the maintainer) :)
https://github.com/coreutils/coreutils/commit/8367b95a13b96b96db93a7339c094f813af877c9
(as author, I can share this link)

Comment (BenWiederhake) [reply src/uu/ls/src/ls.rs:1084]: Ah! Then it's by definition the same behavior. Makes sense :D

Maybe `ls` should warn when `--format=single-column` is ignored due to this? But that's a topic out of scope for this PR.

Comment (sylvestre) [reply src/uu/ls/src/ls.rs:1084]: yeah, i will have a look in a different PR (and maybe add a test on the GNU side)

**Review by BenWiederhake — APPROVED** [review]
Implements and tests exactly what's on the tin. Awesome!

File: src/uu/ls/src/ls.rs [hunk 2]
```diff
@@ -1091,7 +1086,8 @@ impl Config {
...
```
Comment (tertsdiepraam) [inline src/uu/ls/src/ls.rs:1091]: Isn't this if expression not just this?
```rust
if dired {
    format = Format::Long;
}
```

**Review by tertsdiepraam — COMMENTED** [review]
There's a bit of a clap issue again here, I believe (but I don't have 9.5 installed yet so I can't check, I'll verify soon). The problem is that we are forcing the long format after all arguments have been processed by clap, whereas even the new GNU version, does it as soon as `--dired` is encountered. Hence, we'll probably be incompatible in this case:
```
ls --dired --format=vertical
```
Because GNU still overrides it but we don't. Not necessarity a super big problem, but something to be aware of.

File: tests/by-util/test_ls.rs [hunk 3]
```diff
@@ -3749,15 +3749,31 @@ fn test_ls_perm_io_errors() {
...
+    let at = &scene.fixtures;
+    at.mkdir("dir");
+    at.touch("dir/a");
+    scene
+        .ucmd()
+        .arg("--dired")
+        .arg("--hyperlink")
+        .arg("-R")
+        .succeeds()
+        .stdout_contains("  ./dir:");
```
Comment (cakebaker) [inline tests/by-util/test_ls.rs:3776]: To fix the failing Windows tests:
```suggestion
        .stdout_contains(format!("  .{}dir:", std::path::MAIN_SEPARATOR_STR));
```

[the remark this answers is not in the document]
Comment (sylvestre) [reply tests/by-util/test_ls.rs:3956]: One more time some ordering issue:
$ ../gnu/src/ls -R --dired --hyperlink a2    
will show hyperlink
but
$ ../gnu/src/ls -R --hyperlink --dired a2     
won't

[the remark this answers is not in the document]
Comment (sylvestre) [reply tests/by-util/test_ls.rs:3956]: of course, I am responsible for this change upstream ... 
https://github.com/coreutils/coreutils/commit/8367b95a13b96b96db93a7339c094f813af877c9

File: tests/by-util/test_ls.rs [hunk 4]
```diff
@@ -3929,15 +3929,66 @@ fn test_ls_perm_io_errors() {
...
+    at.mkdir("dir");
+    at.touch("dir/a");
+    scene
+        .ucmd()
+        .arg("--dired")
+        .arg("--hyperlink")
+        .arg("-R")
+        .succeeds()
+        .stdout_contains("file://")
+        .stdout_does_not_contain("//DIRED//");
```
Comment (cakebaker) [inline tests/by-util/test_ls.rs:3959]: While this test passes, the output is different from GNU `ls`:
```
$ ls --dired --hyperlink -R ../playground/ls_6144
../playground/ls_6144:
total 4
drwxr-xr-x 2 dho dho 4096 Jun 23 14:56 dir

../playground/ls_6144/dir:
total 0
-rw-r--r-- 1 dho dho 0 Jun 23 14:56 a

$ cargo run ls --dired --hyperlink -R ../playground/ls_6144
../playground/ls_6144:
dir

dir:
a
```
If `--dired` and `--hyperlink` are swapped, both uutils and GNU show the same output.

File: src/uu/ls/src/ls.rs [hunk 5]
```diff
@@ -1084,8 +1079,9 @@ impl Config {
         };
 
         let dired = options.get_flag(options::DIRED);
-        if dired && format != Format::Long {
-            return Err(Box::new(LsError::ConflictingArgumentDired));
+        if dired {
+            // --dired implies --format=long
+            format = Format::Long;
         }
         if dired && format == Format::Long && options.get_flag(options::ZERO) {
```
Comment (cakebaker) [inline src/uu/ls/src/ls.rs:1086]: With the changes on lines 1082-1084 you no longer have to check `format`:
```suggestion
        if dired && options.get_flag(options::ZERO) {
```

File: src/uu/ls/src/ls.rs [hunk 6]
```diff
@@ -1306,7 +1304,7 @@ pub fn uu_app() -> Command {
                 .num_args(0..=1)
                 .default_missing_value("always")
                 .default_value("never")
-                .value_name("WHEN"),
+                .value_name("WHEN").overrides_with(options::DIRED),
```
Comment (cakebaker) [inline src/uu/ls/src/ls.rs:1307]: For some reason `fmt` doesn't work in this function:
```suggestion
                .value_name("WHEN")
                .overrides_with(options::DIRED),
```

File: tests/by-util/test_ls.rs (hunk 4 above)
Comment (sylvestre) [reply tests/by-util/test_ls.rs:3959]: one more time, I am impressed by your review skills, bravo :)

this is probably a small bug on the gnu bug.
hyperlink overrides only part of dired

**Comment by sylvestre** [comment]
I tried with https://docs.rs/clap/latest/clap/struct.ArgMatches.html#method.index_of but it doesn't work with the setTrue

File: src/uu/ls/src/dired.rs [hunk 7]
```diff
@@ -179,6 +179,16 @@ pub fn update_positions(dired: &mut DiredOutput, start: usize, end: usize) {
...
 }
 
+/// Checks if the "--dired" or "-D" argument is present in the command line arguments.
+/// we don't use clap here because we need to know if the argument is present
+/// as it can be overridden by --hyperlink
+pub fn is_dired_arg_present() -> bool {
+    let args: Vec<String> = std::env::args().collect();
+    let dired_index = args.iter().position(|x| x == "--dired" || x == "-D");
+
+    dired_index.is_some()
```
Comment (cakebaker) [inline src/uu/ls/src/dired.rs:189]: As you are only interested in the presence of `--dired`/`-D` but not the position, you can simplify it to:
```suggestion
    args.iter().any(|x| x == "--dired" || x == "-D")
```

**Comment by cakebaker** [comment]
> Congrats! The gnu test tests/ls/dired is no longer failing!

Great, good job :)

**Pull request merged**